text/0000-relax-mixin-contract.md

```diff
@@ -0,0 +1,69 @@
+# Meta
+[meta]: #meta
+- Name: Relax Mixin Contract
+- Start Date: 2020-08-12
+- Author(s): @sclevine
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Mixins are used to imply multiple contracts:
+
+1. A contract between a buildpack and a stack, where the buildpack may state that it can only be used with stack images that have certain dependencies.
+
+2. A contract between a stack run image and stack build image, where apps built on the build image must have a certain dependency present on the run image to start successfully on the run image.
+
+Currently, if a stack build image has the mixin `build:mypkg` and a stack run image has the mixin `run:mypkg`, a buildpack cannot require the mixin as `mypkg` (either statically in buildpack.toml, or dynamically via the stackpack mechanism described in #87). A stack author may choose to use this configuration because the same dependency is used in different, unrelated ways in both images.
+
+Similarly, if a stack build image shares the mixin `mypkg` with the stack run image, a buildpack cannot require the mixin `build:mypkg` or `run:mypkg`, even though both images contain `mypkg`.
+
+This RFC proposed that we update the buildpack specification to create a new relationship between `mypkg`, `run:mypkg`, and `build:mypkg`.
+
+- While `run:mypkg` may imply a different set of changes to the run image compared to `mypkg`, buildpacks requiring `mypkg` should be satisfied with `run:mypkg` at runtime (and vice versa).
+
+- While `build:mypkg` may imply a different set of changes to the build image compared to `mypkg`, buildpacks requiring `mypkg` should be satisfied with `build:mypkg` at build-time (and vice versa).
+
+- A runtime stackpack would still be passed either `run:mypkg` or `mypkg` depending on what the buildpack required, so that the stackpack understands whether or not it can expect the build image to have the corresponding mixin as well.
```

Comment:
Can you give an example of a case where the difference between `run:mypkg` and `mypkg` matters to the stackpack (while it is installing packages  on the run-image)?

Reply:
Example: requesting `build:libfoo` could mean adding just libfoo.so at build time for use by build tools, but requesting `libfoo` might mean adding both libfoo.so and libfoo.h at build time, so that apps can link against libfoo.so during build and expect it to be available at launch

Reply:
Do we think users want that level of granularity in mixins? I.E. is there a good reason a user would definitely NOT want `libfoo.h` at build time? Sometimes getting headers when you don't need them might be worth it for the sake of simplicity in the contract? I would be less willing to make this argument given a realistic runtime example. Accidentally requiring anything extra at runtime seems worse than extras at build time.

Also, b/c stack authors can define the meaning of a mixin name, they could define two mixins: one that means `libfoo.so` and another that includes the headers (like the bionic mixin names). I think we should let folks build these distinctions into the mixin definitions themselves without specified phase prefixes.

Reply:
I don't think this matters very much, practically. But it would be a more drastic change if `run:mypkg` and `mypkg` were considered the same mixin, and `run:` was just a property the buildpack/stack used to imply where the mixin must be installed. I'm worried that there are cases where the stack<->stack contract and buildpack<->stack contract are interwoven, so that the stackpack needs all of the information about where the mixin will be installed.

But I could probably be convinced to switch to this approach if you felt strongly that the iterative changes described are worse than the current state.

Reply:
Switched as discussed.